File: 01-calorie-counting/rust/src/main.rs
use utils::{self, fs::DataType};

fn main() {
    let contents = utils::fs::read_data(DataType::Real);
    let highest = part_one(&contents);
    let top_three = part_two(&contents);

    println!("First solution: {highest:#?}");
    println!("Second solution: {top_three}");
}

fn part_one(contents: &str) -> usize {

    let elves: Vec<String> = contents
        .split("\n\n")
        .map(String::from)
        .collect();

    let mut highest = 0;
    let (mut second, mut third) = (0, 0);

    for elf in elves {
        let calories: usize = elf
            .split('\n')
            .map(|s| str::parse(s).unwrap())
            .reduce(|accum, item| accum + item)
            .unwrap();

        if calories >= highest {
            third = second;
            second = highest;
            highest = calories;
        } else if calories >= second {
            third = second;
            second = calories;
        } else if calories > third {
            third = calories;
        }
    }

    highest
}

fn part_two(contents: &str) -> usize {
    let elves: Vec<String> = contents
        .split("\n\n")
        .map(String::from)
        .collect();

    let mut highest = 0;
    let (mut second, mut third) = (0, 0);

    for elf in elves {
        let calories: usize = elf
            .split('\n')
            .map(|s| str::parse(s).unwrap())
            .reduce(|accum, item| accum + item)
            .unwrap();

        if calories >= highest {
            third = second;
            second = highest;
            highest = calories;
        } else if calories >= second {
            third = second;
            second = calories;
        } else if calories > third {
            third = calories;
        }
    }

    highest + second + third
}

#[cfg(test)]
mod tests {
    use super::{part_one, part_two};

    const INPUT: &str = "1000
2000
3000

4000

5000
6000

7000
8000
9000

10000";

    #[test]
    fn part_one_works() {
        let expected = 24000;
        let actual = part_one(INPUT);

        assert_eq!(actual, expected);
    }

    #[test]
    fn part_two_works() {
        let expected = 45000;
        let actual = part_two(INPUT);

        assert_eq!(actual, expected);
    }
}

File: 01-calorie-counting/go/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"

	"lazytangent/aoc/2022/utils"
)

func main() {
	input, err := utils.ParseInput(os.Args)
	if err != nil {
		log.Fatal(err)
	}

	input = strings.TrimSpace(input)
	elves := strings.Split(input, "\n\n")

	max := 0
	second, third := 0, 0
	for _, elf := range elves {
		curr := 0
		for _, item := range strings.Split(elf, "\n") {
			val, _ := strconv.Atoi(item)
			curr += val
		}

		if curr > max {
			third = second
			second = max
			max = curr
		} else if curr > second {
			third = second
			second = curr
		} else if curr > third {
			third = curr
		}
	}

	fmt.Printf("First solution: %d\n", max)
	fmt.Printf("Second solution: %d\n", max + second + third)
}
